### PR TITLE
genericapiserver: cut off more dependencies – episode 4

### PR DESCRIPTION
**Pull request by sttts**
Follow-up of https://github.com/kubernetes/kubernetes/pull/40426.

TODO:
- [x] resync client-go before "genericapiserver: cutting off pkg/api deps" when #40426 went in and mirror repos are synched.

approved based on #40363

### CODE REVIEW

File: pkg/genericapiserver/endpoints/filters/impersonation.go [hunk 1]
```diff
@@ -29,8 +29,8 @@ import (
 	"k8s.io/apiserver/pkg/authorization/authorizer"
 	"k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/server/httplog"
-	"k8s.io/kubernetes/pkg/api"
-	authenticationapi "k8s.io/kubernetes/pkg/apis/authentication"
+	"k8s.io/client-go/pkg/api"
+	authenticationapi "k8s.io/client-go/pkg/apis/authentication"
```
Comment (deads2k) [inline pkg/genericapiserver/endpoints/filters/impersonation.go]: this definitely makes the dependency tree weird.  We should have issue to deal with it.  server shouldn't trust client for its impersonation headers.  I think I'm actually in favor of a copy for the short term instead.

File: pkg/genericapiserver/endpoints/filters/impersonation.go [hunk 2]
```diff
@@ -29,8 +29,8 @@ import (
 	"k8s.io/apiserver/pkg/authorization/authorizer"
 	"k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/server/httplog"
-	"k8s.io/kubernetes/pkg/api"
-	authenticationapi "k8s.io/kubernetes/pkg/apis/authentication"
+	"k8s.io/client-go/pkg/api"
```
Comment (deads2k) [inline pkg/genericapiserver/endpoints/filters/impersonation.go]: delete and build the groupkinds by hand

File: pkg/genericapiserver/registry/rest/update.go [hunk 3]
```diff
@@ -143,7 +142,7 @@ func DefaultUpdatedObjectInfo(obj runtime.Object, copier runtime.ObjectCopier, t
 }
 
 // Preconditions satisfies the UpdatedObjectInfo interface.
-func (i *defaultUpdatedObjectInfo) Preconditions() *api.Preconditions {
+func (i *defaultUpdatedObjectInfo) Preconditions() *metav1.Preconditions {
```
Comment (deads2k) [inline pkg/genericapiserver/registry/rest/update.go]: these were just straight up wrong before?

File: pkg/genericapiserver/endpoints/filters/impersonation.go (hunk 1 above)
Comment (sttts) [inline pkg/genericapiserver/endpoints/filters/impersonation.go]: 100% agree. I don't like it. We should have a test group in k8s.io/apiserver and use that everywhere.

**Comment by deads2k** [comment]
```
pkg/api: move HasObjectMetaSystemFieldValues into apimachinery - fine
pkg/genericapiserver: move resttest to pkg/registry - not sold.  I think we want this to be something people get in the apiserver library.  Maybe just pass in a scheme and a semantic
```

File: pkg/genericapiserver/registry/rest/update.go (hunk 3 above)
Comment (sttts) [inline pkg/genericapiserver/registry/rest/update.go]: at least forgotten during metav1 refactors.

Comment (sttts) [inline pkg/genericapiserver/registry/rest/update.go]: @smarterclayton any plan to remove the deprecated types like these in kube?

**Comment by deads2k** [comment]
pkg/genericapiserver: move NewDefaultAPIGroupInfo to pkg/registry - disagree. I think we should accept default schemes, codecs, and registries when we start and keep the helper.  That will give one initialization point for people and anyone sane only has one.

**Comment by deads2k** [comment]
pkg/genericapiserver/server: cut off pkg/api dependencies from test - lgtm

File: pkg/genericapiserver/endpoints/installer.go [hunk 4]
```diff
@@ -154,8 +153,9 @@ func (a *APIInstaller) getResourceKind(path string, storage rest.Storage) (schem
 			break
 		}
 
-		// TODO This keeps it doing what it was doing before, but it doesn't feel right.
-		if fqKind.Group == extensions.GroupName && fqKind.Kind == "ThirdPartyResourceData" {
+		// TODO: keep rid of extensions api group dependency here
+		// This keeps it doing what it was doing before, but it doesn't feel right.
+		if fqKind.Group == "extensions" && fqKind.Kind == "ThirdPartyResourceData" {
```
Comment (deads2k) [inline pkg/genericapiserver/endpoints/installer.go]: blech.  I'm so sad this leaks into generic.

**Comment by deads2k** [comment]
pkg/genericapiserver/endpoints: cut off extensions api dep - lgtm

**Comment by deads2k** [comment]
pkg/storage/storagebackend/factory: cut off from testapi - lgtm

File: pkg/storage/etcd3/store_test.go [hunk 5]
```diff
@@ -22,17 +22,18 @@ import (
 	"sync"
 	"testing"
 
+	apitesting "k8s.io/apimachinery/pkg/api/testing"
```
Comment (deads2k) [inline pkg/storage/etcd3/store_test.go]: I don't recognize this package

File: pkg/genericapiserver/endpoints/installer.go (hunk 4 above)
Comment (sttts) [inline pkg/genericapiserver/endpoints/installer.go]: its ugliness is highly visible now. This is good!

File: pkg/storage/etcd3/store_test.go (hunk 5 above)
Comment (deads2k) [inline pkg/storage/etcd3/store_test.go]: Oh, this was the fuzzer and stuff wasn't it

File: staging/src/k8s.io/apimachinery/pkg/api/testing/register.go [hunk 6]
```diff
@@ -25,10 +25,12 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	runtimeserializer "k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientapi "k8s.io/client-go/pkg/api"
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/register.go]: no.  No cycles

File: pkg/storage/etcd3/store_test.go (hunk 5 above)
Comment (sttts) [inline pkg/storage/etcd3/store_test.go]: it's where the fuzzer went. Am undecided yet whether it's good or bad: it's where I create a scheme, codecs, etc. for tests, registering meta types and client core types. The later should be replaced with real test group types in a follow-up.

File: pkg/storage/storagebackend/factory/tls_test.go [hunk 7]
```diff
@@ -67,7 +57,7 @@ func TestTLSConnection(t *testing.T) {
 		CertFile:   certFile,
 		KeyFile:    keyFile,
 		CAFile:     caFile,
-		Codec:      codec,
+		Codec:      apitesting.Codec(schema.GroupVersion{Version: "v1"}),
```
Comment (deads2k) [inline pkg/storage/storagebackend/factory/tls_test.go]: you should probably create your own scheme and codecfactory in these packages.

**Comment by deads2k** [comment]
pkg/storage/etcd3: cut off from testapi - introduces a cycle

**Comment by deads2k** [comment]
`staging/copy.sh: use compatible 'sed -i=""' everywhere` do not want.  Is this a "gee, I wish mac OS was compatible" thing?

File: staging/copy.sh [hunk 8]
```diff
@@ -102,10 +102,9 @@ find "${MAIN_REPO}/pkg/version" -maxdepth 1 -type f | xargs -I{} cp {} "${CLIENT
 # need to copy clientsets, though later we should copy APIs and later generate clientsets
 mkcp "pkg/client/clientset_generated/${CLIENTSET}" "pkg/client/clientset_generated"
 mkcp "/pkg/client/record" "/pkg/client"
-
 mkcp "/pkg/client/unversioned/portforward" "/pkg/client/unversioned"
-
 mkcp "/pkg/util/workqueue" "pkg/util"
+mkcp "/pkg/api/testing" "pkg/api"
```
Comment (deads2k) [inline staging/copy.sh]: this is moving us in the wrong direction.

Comment (sttts) [inline staging/copy.sh]: We could use gsed instead in fact, forcing Mac users into installing it (which is acceptible IMO).

File: pkg/storage/etcd3/store_test.go (hunk 5 above)
Comment (deads2k) [inline pkg/storage/etcd3/store_test.go]: > it's where the fuzzer went. Am undecided yet whether it's good or bad: it's where I create a scheme, codecs, etc. for tests, registering meta types and client core types. The later should be replaced with real test group types in a follow-up.

I don't want a global testing scheme, codec, registry.  I think that worked out badly.  Helpers to make it easy to create a local one are fine and re-using a real one (like from client-go) is also fine.

**Comment by sttts** [comment]
Reduced to lgtm'ed commits. Applying the label.

/lgtm

**Comment by sttts** [comment]
@k8s-bot kops aws e2e test this

**Pull request merged**